Add hardcoded strings from BlockingError.jsx

treyhahn opened the pull request:
These are strings related to error messages and buttons from the BlockingError.jsx file.

treyhahn commented:
This includes the new strings from https://github.com/streetmix/streetmix/pull/835/files#diff-3baa06db0e75f730d7a02f06eb95228cR72.

@louh, how is the organization of this added code? I created a new category called error-msg, but want to make sure it isn't cluttered or unorganized. How can I improve it?

Also, for the following string it doesn't look like there is anything to translate. Do I still include the description as a string? This is not included in my PR.
https://github.com/streetmix/streetmix/blob/28ef7fa80101307bbf5212e3a1ace0bf0d0cdd1d/assets/scripts/app/BlockingError.jsx#L63-L66

Finally, there are some duplicate title strings. Do we need all of them? Will they show up twice in Transifex?
![image](https://user-images.githubusercontent.com/12552044/37219338-10ac587e-2391-11e8-9417-6be774310805.png)

louh commented:
More important than the organization, it's the syntax. You're working in [JSON format](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/JSON), which can be very unforgiving when it comes to things like missing commas. I would recommend taking the contents of the `main.json` file and pasting it into https://jsonlint.com/ to see where your syntax errors are, and fixing them one by one. Another way to see JSON parse errors is by looking at the file diff in this pull request (https://github.com/streetmix/streetmix/pull/842/files) and seeing where it's highlighted in red.

In particular, you will want to pay attention to the following:

### Missing commas

Within an object, key-value pairs must be separated by a comma, placed at the end of each key-value pair, _except_ for the last one in a series.

**NO:**

```json
{
  "foo": "bar"
  "baz": "qux"
}
```

Notice the highlighted red `:` because the missing comma is causing the semicolon to be an unexpected syntax character.

**YES:**

```json
{
  "foo": "bar",
  "baz": "qux"
}
```

### Extra trailing commas

Within an object, the _last_ key-value pairs must not be followed by a comma.

**NO:**

```json
{
  "foo": "bar",
  "baz": "qux",
}
```

**YES:**

```json
{
  "foo": "bar",
  "baz": "qux"
}
```

### No line breaks within values

In a value, there must not be a line break.

**NO:**

```json
{
  "foo": "bar
baz
  "
}
```

**YES:**

```json
{
  "foo": "bar baz"
}
```

### Double quotation marks within values require special treatment.

A value is surrounded by `"double quotation marks"`. What happens if you try to use a double quote inside a value, `"for instance, "something" like this"`? Unfortunately this will break because when the next double quote is encountered, the parser will treat it as the end of the value. In other words, in this last example, the value is `"for instance, "` and everything that follows is a syntax error.

**NO:**

```json
{
  "foo": "bar "baz" qux"
}
```

**YES (1):**

One way to solve this is to switch to single quotation marks inside the value.

```json
{
  "foo": "bar 'baz' qux"
}
```

**YES (2):**

If you absolutely require double quotation marks inside the value, you can "escape" the value with a backslash `\`:

```json
{
  "foo": "bar \"baz\" qux"
}
```

---

Fix these syntax errors first, then we will talk about organization.

treyhahn commented:
Thank you for the extremely thorough feedback @louh! I think I fixed all the syntax errors.

louh commented:
Thanks Trey!

Now, let's do some cleanup.

- We want to keep the key names short but descriptive. I think in this case the top level key `error-msg` can be replaced with `errors` which is a little shorter, easier to type, and there's no loss in meaning.
- With the repeated strings, e.g. "Having trouble" or "Street not found" they share a common characteristic: they're generic labels for more specific error messages. I don't think we need to duplicate them here, because yes, they will be duplicated in Transifex as different strings. Consolidating them to a single string that serves multiple error types is just fine.

treyhahn commented:
One Question: there are slight variations of the `We're having trouble loading Streetmix.` string that already exists: https://github.com/streetmix/streetmix/blob/cea0f596a60c7c73acf95b415d899937591588a0/assets/locales/en/main.json#L61

Should we add the strings in the screenshot below?
![image](https://user-images.githubusercontent.com/12552044/37300482-769cfad6-25fc-11e8-9be4-c2890bc3f58f.png)

treyhahn commented:
The key has been changed to `errors` and repeated strings have been deleted.

A note: there is already a `"Having trouble..."` string in the `msg` key, so I removed all `"Having trouble..."`'s in the `errors` key. Same with `No street selected.`.
https://github.com/streetmix/streetmix/blob/cea0f596a60c7c73acf95b415d899937591588a0/assets/locales/en/main.json#L59-L60

louh commented:
Re: `We're having trouble loading Streetmix`, let's just have one string of
just this text. For the second part of that message, we'll have it split
into a second string: `(Error {{code}}.)`



On Mon, Mar 12, 2018, 1:55 PM Trey Hahn <notifications@github.com> wrote:

> The key has been changed to errors and repeated strings have been deleted.
>
> A note: there is already a "Having trouble..." string in the msg key, so
> I removed all "Having trouble..."'s in the errors key. Same with No
> street selected..
>
> https://github.com/streetmix/streetmix/blob/cea0f596a60c7c73acf95b415d899937591588a0/assets/locales/en/main.json#L59-L60
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/streetmix/streetmix/pull/842#issuecomment-372404459>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACb1tGKNaWS9Usc9Wu7Xcm8TzSVBVs3Oks5tdraNgaJpZM4Skjpw>
> .
>

treyhahn commented:
Copy. Any further edits?

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+			"sign-in-button": "Sign in again",
+			"reload-button": "Reload the page",
+			"try-again-button": "Try again",
+			"general-help-buttons": "<a target='_blank' href='mailto:hello@streetmix.net'>email</a> or <a target='_blank' href='https://twitter.com/intent/tweet?text=@streetmix'>Twitter</a>. <br /> <button id = 'error-home'onClick = {goHome} > Go to the homepage < /button>"
```
For these buttons, let's remove the `-buttons` suffix, as it's redundant. The parent label is already `buttons`.

For `general-help-buttons` (soon to be renamed to `general-help`), I think we should remove this string. I'll explain further below.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+			"general-help-buttons": "<a target='_blank' href='mailto:hello@streetmix.net'>email</a> or <a target='_blank' href='https://twitter.com/intent/tweet?text=@streetmix'>Twitter</a>. <br /> <button id = 'error-home'onClick = {goHome} > Go to the homepage < /button>"
+		},
+		"page-not-found-title": "Page not found.",
+		"page-not-found-description": "Oh, boy. There is no page with this address!<br />{homeButton}",
```
Let's remove everything after, and including, the `<br />`. We can handle that part in React, we don't need to do this in the translation string.

treyhahn replied:
Ok great, I only included it since I don't understand how the string translations get connected to the actual UI.

louh replied:
No worries. This is a normal process of trying something and thinking about whether it works or not.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+		"page-not-found-title": "Page not found.",
+		"page-not-found-description": "Oh, boy. There is no page with this address!<br />{homeButton}",
+		"street-not-found-title": "Street not found.",
+		"street-not-found-description": "Oh, boy. There is no street with this link!<br />{homeButton}",
```
Remove everything after, and including `<br />`.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+		"page-not-found-description": "Oh, boy. There is no page with this address!<br />{homeButton}",
+		"street-not-found-title": "Street not found.",
+		"street-not-found-description": "Oh, boy. There is no street with this link!<br />{homeButton}",
+		"street-not-found-but-user-description": "There is no street with this link! But you can look at other streets by {linkToUser(this.props.street)} <br />{homeButton}",
```
Remove everything after, and including `<br />`.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+		"street-not-found-description": "Oh, boy. There is no street with this link!<br />{homeButton}",
+		"street-not-found-but-user-description": "There is no street with this link! But you can look at other streets by {linkToUser(this.props.street)} <br />{homeButton}",
+		"street-deleted-title": "This street has been deleted.",
+		"street-deleted-description": "There is no longer a street with this link, but you can look at other streets by {linkToUser(this.props.street)} <br />{homeButton}",
```
Remove everything after, and including `<br />`.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+		"sign-out-title": "You are now signed out.",
+		"reload-sign-out-title": "You signed out in another window.",
+		"gallery-street-failure-description": "We’re having trouble loading this street.",
+		"reload-sign-out-description-401": "Please reload this page before continuing.<br />(Error RM2.)<br /><button id='error-clear-sign-in-reload' onClick={goReloadClearSignIn}>Reload the page</button>",
```
Remove everything after, and including `<br />` (the first one). Notice that we can handle the middle string with the `error-code` string and the latter part within React and the `error.button.reload` string.

treyhahn replied:
Ahhh, this is great for me to learn. Now I'm starting to understand things a bit more.

treyhahn replied:
@louh This question may be too much for now, but how are these strings brought to the actual  UI? Where in the code does this happen. Is someone having to look at this translation file and reference all of these different keys we are creating?

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
     "slow-connection-2": "You might want to reload the page and try again. Please note you might lose the latest change to the street. Sorry!",
     "no-connection": "Streetmix is having trouble connecting to the Internet."
+  },
+	"errors": {
```
In keeping with the syntax elsewhere, we should use the singular form, so `error` instead of `errors` (I got this wrong last time).

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
     "no-connection": "Streetmix is having trouble connecting to the Internet."
+  },
+	"errors": {
+		"buttons": {
```
Similarly, singular form here -- `button` instead of `buttons`.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+		"gallery-street-failure-description": "We’re having trouble loading this street.",
+		"reload-sign-out-description-401": "Please reload this page before continuing.<br />(Error RM2.)<br /><button id='error-clear-sign-in-reload' onClick={goReloadClearSignIn}>Reload the page</button>",
+		"reload-sign-in-title": "You signed in in another window.",
+		"reload-sign-in-description": "Please reload this page before continuing.<br />{reloadButton}",
```
Remove everything after, and including `<br />`.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+		"reload-sign-in-title": "You signed in in another window.",
+		"reload-sign-in-description": "Please reload this page before continuing.<br />{reloadButton}",
+		"street-deleted-elsewhere-title": "This street has been deleted elsewhere.",
+		"street-deleted-elsewhere-description": "This street has been deleted in another browser.<br />{homeButton}",
```
Remove everything after, and including `<br />`.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+		"street-deleted-elsewhere-title": "This street has been deleted elsewhere.",
+		"street-deleted-elsewhere-description": "This street has been deleted in another browser.<br />{homeButton}",
+		"twitter-access-denied-title": "You are not signed in.",
+		"twitter-access-denied-description": "You cancelled the Twitter sign in process.<br />{homeButton}",
```
Remove everything after, and including `<br />`.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+		"unsupported-browser-title": "Streetmix doesn’t work on your browser.",
+		"unsupported-browser-description": "Sorry about that. You might want to try <a target='_blank' href='http://www.google.com/chrome'>Chrome</a>,&nbsp;<a target='_blank' href='http://www.mozilla.org/firefox'>Firefox</a>,&nbsp;<a target='_blank' href='https://www.microsoft.com/en-us/windows/microsoft-edge'>Microsoft Edge</a>, or Safari.<br /><br />Are you on Internet Explorer?&nbsp;<a target='_blan ...
+		"cannot-create-new-street-on-phone-title": "Streetmix works on tablets and desktops only.",
+		"cannot-create-new-street-on-phone-description": "If you follow another link to a specific street, you can view it on your phone – but you cannot yet create new streets.<br /> <button id='error-example' onClick={goExampleStreet}>View an example street</button>",
```
Following similar examples above, let's move "View an example street" as a separate string and put it under `error.button`. Then remove remove everything after, and including `<br />` from this string.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+		"auth-api-problem-title": "There was a problem with signing you in.",
+		"auth-api-problem-description": "There was a problem with Twitter authentication. Please try again later or let us know via {generalHelpButtons}",
+		"unsupported-browser-title": "Streetmix doesn’t work on your browser.",
+		"unsupported-browser-description": "Sorry about that. You might want to try <a target='_blank' href='http://www.google.com/chrome'>Chrome</a>,&nbsp;<a target='_blank' href='http://www.mozilla.org/firefox'>Firefox</a>,&nbsp;<a target='_blank' href='https://www.microsoft.com/en-us/windows/microsoft-edge'>Microsoft Edge</a>, or Safari.<br /><br />Are you on Internet Explorer?&nbsp;<a target='_blan ...
```
This is very long and includes a lot of HTML markup, so I think we could make translators' lives easier (and future code maintainer's lives easier) but splitting this description in 3 parts.

1. The "try another browser" sentence.
2. The "Internet Explorer" sentence. This could even be split into 2 more parts, the label and the text link.
3. The "contact us" sentence.

For ease of maintainance, replace links and email (anything inside the `href=` attribute) with a string-replacement placeholder, e.g. `{chromeLink}`.

treyhahn replied:
@louh Got it. HTML is already handled by Transifex like this:
![image](https://user-images.githubusercontent.com/12552044/37413685-f821f84a-277d-11e8-9a6d-e56b3bd99436.png)

We could remove it or not. Removing it would create less of those placeholder numbers (the string in screenshot above has a lot of them). How would it affect the code though down the line? Would it help us to keep it in?

louh replied:
This really depends. Let's talk about this tomorrow.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+		"twitter-access-denied-title": "You are not signed in.",
+		"twitter-access-denied-description": "You cancelled the Twitter sign in process.<br />{homeButton}",
+		"auth-api-problem-title": "There was a problem with signing you in.",
+		"auth-api-problem-description": "There was a problem with Twitter authentication. Please try again later or let us know via {generalHelpButtons}",
```
Let's break up the "Please try again later" message into a new string and instead of using `{generalHelpButtons}` let's bake that string right in to the message, like so:

```
Please try again later or let us know via <a target='_blank' href='mailto:hello@streetmix.net'>email</a> or <a target='_blank' href='https://twitter.com/intent/tweet?text=@streetmix'>Twitter</a>.
```

To make our lives easier, we should replace the URLs with placeholders.

```
Please try again later or let us know via <a target='_blank' href='{email}'>email</a> or <a target='_blank' href='{twitter}'>Twitter</a>.
```

treyhahn replied:
Do we want to keep the "Please try again later" in the `error` key or put it in the `msg` key? Other question here, should we remove the URLs or just keep them there since Transifex already detects them?

treyhahn replied:
And then should we use {} or {{}} as placeholders? On the Transifex end, should we tell it to detect both for in Transifex UI?

louh replied:
I'll try to keep things simple.

1. We will _eventually_ use single braces `{}`.
2. We _currently_ use double braces `{{}}`.

Let's _currently_ be consistent and use double braces for everything.

When we are ready to switch to single braces, then we will adjust the strings. Although, I'm not sure if adjusting the strings will erase the old translations, so let's be aware that that might be the case.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,44 @@
+		"cannot-create-new-street-on-phone-title": "Streetmix works on tablets and desktops only.",
+		"cannot-create-new-street-on-phone-description": "If you follow another link to a specific street, you can view it on your phone – but you cannot yet create new streets.<br /> <button id='error-example' onClick={goExampleStreet}>View an example street</button>",
+		"generic-error-title": "Something went wrong.",
+		"generic-error-description": "We’re sorry – something went wrong. Please try again later or let us know via {generalHelpButtons}"
```
You can probably see where I was going with this when I said to break out the second sentence from `auth-api-problem-description`. The new string we created from that one can be used here, so we can remove the second sentence from `generic-error-description`.

louh commented:
Now we refactor!

louh commented:
Yep!

I'm thinking tomorrow when we meet in person we spend some time working on
bringing the translation text into the UI. I'll do the typing, but I think
when you see me put it together, you'll really get to see what's happening.

On Wed, Mar 14, 2018 at 11:39 AM, Trey Hahn <notifications@github.com>
wrote:

> *@treyhahn* commented on this pull request.
> ------------------------------
>
> In assets/locales/en/main.json
> <https://github.com/streetmix/streetmix/pull/842#discussion_r174504974>:
>
> > +			"sign-in-button": "Sign in again",
> +			"reload-button": "Reload the page",
> +			"try-again-button": "Try again",
> +			"general-help-buttons": "<a target='_blank' href='mailto:hello@streetmix.net'>email</a> or <a target='_blank' href='https://twitter.com/intent/tweet?text=@streetmix'>Twitter</a>. <br /> <button id = 'error-home'onClick = {goHome} > Go to the homepage < /button>"
> +		},
> +		"page-not-found-title": "Page not found.",
> +		"page-not-found-description": "Oh, boy. There is no page with this address!<br />{homeButton}",
> +		"street-not-found-title": "Street not found.",
> +		"street-not-found-description": "Oh, boy. There is no street with this link!<br />{homeButton}",
> +		"street-not-found-but-user-description": "There is no street with this link! But you can look at other streets by {linkToUser(this.props.street)} <br />{homeButton}",
> +		"street-deleted-title": "This street has been deleted.",
> +		"street-deleted-description": "There is no longer a street with this link, but you can look at other streets by {linkToUser(this.props.street)} <br />{homeButton}",
> +		"sign-out-title": "You are now signed out.",
> +		"reload-sign-out-title": "You signed out in another window.",
> +		"gallery-street-failure-description": "We’re having trouble loading this street.",
> +		"reload-sign-out-description-401": "Please reload this page before continuing.<br />(Error RM2.)<br /><button id='error-clear-sign-in-reload' onClick={goReloadClearSignIn}>Reload the page</button>",
>
> @louh <https://github.com/louh> This question may be too much for now,
> but how are these strings brought to the actual UI? Where in the code does
> this happen. Is someone having to look at this translation file and
> reference all of these different keys we are creating?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/streetmix/streetmix/pull/842#discussion_r174504974>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACb1tNZbbb7ZwvqljeHOpn6SX1U-Dmj0ks5teTmTgaJpZM4Skjpw>
> .
>

treyhahn commented:
Sounds good! I made the changes except where I had additional questions/comments. I left those on your review comments.

Also, what do you mean by refactor? Is this changing the structure of the code?

louh commented on assets/locales/en/main.json:
```diff
@@ -64,39 +64,42 @@
+			"sign-in": "Sign in again",
+			"reload": "Reload the page",
+			"try-again": "Try again",
+      "view-example": "View an example street"
```
Extra space indent here. It doesn't affect the validity of the JSON file, but I am a stickler for detail.

louh replied:
Actually, I think I know what's going on here. What are you using to edit this? It's using tabs instead of spaces, and we should be consistent and use only spaces for indents.

louh commented on assets/locales/en/main.json:
```diff
@@ -64,39 +64,42 @@
-		"street-not-found-description": "Oh, boy. There is no street with this link!<br />{homeButton}",
-		"street-not-found-but-user-description": "There is no street with this link! But you can look at other streets by {linkToUser(this.props.street)} <br />{homeButton}",
+		"street-not-found-description": "Oh, boy. There is no street with this link!",
+		"street-not-found-but-user-description": "There is no street with this link! But you can look at other streets by {linkToUser(this.props.street)}",
```
It looks like that the value inside the placeholder was originally a JavaScript function, which we won't need here. I would simplify this to `{user}`.

louh commented on assets/locales/en/main.json:
```diff
@@ -64,39 +64,42 @@
+		"street-not-found-but-user-description": "There is no street with this link! But you can look at other streets by {linkToUser(this.props.street)}",
 		"street-deleted-title": "This street has been deleted.",
-		"street-deleted-description": "There is no longer a street with this link, but you can look at other streets by {linkToUser(this.props.street)} <br />{homeButton}",
+		"street-deleted-description": "There is no longer a street with this link, but you can look at other streets by {linkToUser(this.props.street)}",
```
Similar comment to line 79, above.

louh commented on assets/locales/en/main.json:
```diff
@@ -64,39 +64,42 @@
+		"reload-sign-out-description-401": "Please reload this page before continuing.",
 		"reload-sign-in-title": "You signed in in another window.",
-		"reload-sign-in-description": "Please reload this page before continuing.<br />{reloadButton}",
+		"reload-sign-in-description": "Please reload this page before continuing.",
```
This appears to be the same message as on line 85, so let's consolidate it into one.

louh replied:
Let's rename this key to `"please-reload"`.

treyhahn commented:
Edits made. Any next steps? When you said refactor, what does that mean here?

louh commented:
By "refactor" I just meant we're moving things around.

louh commented on assets/locales/en/main.json:
```diff
@@ -59,9 +59,46 @@
     "having-trouble": "Having trouble…",
     "no-street": "No street selected.",
     "trouble-loading": "We’re having trouble loading Streetmix.",
+    "error-code": "(Error {{code}}.)",
```
Should this be moved down to the "error" area with the rest?

treyhahn replied:
Sure, I don't see why not. I included it up there since that seemed to be where the strings that repeated multiple times were.

treyhahn commented:
Should be good to go now. Thanks for your patience with me on this Lou.

louh commented:
@treyhahn Not quite -- please use two-space indents per nested level on these lines: https://github.com/streetmix/streetmix/blob/11caaa4a31b6e282fab831af1590552c9eab801a/assets/locales/en/main.json#L66-L102

treyhahn commented:
Ok, it should hopefully be good now. I also added Arabic to i18n- is that ok to do here in this PR?

louh commented:
It's usually better to keep different "topics" in separate PRs, even if they touch the same file. This is because whether or not Arabic is in the language dropdown isn't really dependent on whether or not error strings are translated -- they're totally different topics. By splitting these into separate PRs you seperate concerns; it means that if error translations end up being a very difficult, long drawn out thing, it doesn't block Arabic from showing up in the language dropdown.

For this one it's going to be more trouble than it's worth to revert the commit and split it up, but just keep that in mind for future reference.

We are close, but I want to open a new branch on top of this one that incorporates the strings into the BlockingError component, so we can continue to verify that the way we split up strings here makes sense. I'd like to do this _before_ we push these strings out to Transifex, otherwise you end up with people having to go back and redo things.

louh commented on assets/locales/en/main.json:
```diff
@@ -63,6 +63,43 @@
+    "cannot-create-new-street-on-phone-description": "If you follow another link to a specific street, you can view it on your phone – but you cannot yet create new streets.",
+    "generic-error-title": "Something went wrong.",
+    "generic-error-description": "We’re sorry – something went wrong.",
+    "error-code": "(Error {{code}}.)"
```
One more change here: use the `{singleBrackets}` here.

treyhahn commented:
Ok, that makes sense. Could I just do a new commit to delete the addition of Arabic and then create a new short PR for it? I don't think that would take that long, no?

Sounds good re: new branch. Let me know if there is further action I can take on it. 

> On Mar 19, 2018, at 5:35 PM, Lou Huang <notifications@github.com> wrote:
> 
> It's usually better to keep different "topics" in separate PRs, even if they touch the same file. This is because whether or not Arabic is in the language dropdown isn't really dependent on whether or not error strings are translated -- they're totally different topics. By splitting these into separate PRs you seperate concerns; it means that if error translations end up being a very difficult, long drawn out thing, it doesn't block Arabic from showing up in the language dropdown.
> 
> For this one it's going to be more trouble than it's worth to revert the commit and split it up, but just keep that in mind for future reference.
> 
> We are close, but I want to open a new branch on top of this one that incorporates the strings into the BlockingError component, so we can continue to verify that the way we split up strings here makes sense. I'd like to do this before we push these strings out to Transifex, otherwise you end up with people having to go back and redo things.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

louh commented:
Don't worry about it, this is very close. I'll merge it tonight.
